### DISCUSSION TITLE
WASM-specific Package for `bw2calc` (on emscripten-forge)

### DISCUSSION THREAD

**Discussion by michaelweinold** [brightway-lca/brightway-live#27] [General]
With permission from @m-rossi, I'm moving our email discussion (between @cmutel, @m-rossi and @michaelweinold) here:

**Comment by michaelweinold** [comment]
`(email by @michaelweinold to @m-rossi and @cmutel from 23.08.2023)`

I tried to add the bw2calc package to emscripten-forge today (cf. [PR#616](https://github.com/emscripten-forge/recipes/pull/616)), but I am still trying to solve some empack-related build issue with two QuantStack developers.
 
As part of this discussion, it was again suggested that we host a noarch version of bw2calc on conda-forge instead of emscripten-forge.
 
How realistic would it be to add a noarch version of bw2calc to the dev label on conda-forge?
 
I believe we have two options:
 
Add a dev version without any LinAlg solver dependencies. I discussed this with @Christopher Mutel who agrees that ‘people who are able to get the dev version from conda-forge can be bothered to install whatever LinAlg solver is suitable for their system’.
Figure out how to have a noarch package where some dependencies are not noarch. This is evidently possible (https://github.com/conda-forge/bw2calc-feedstock/pull/14#issuecomment-1687811107). I would be happy to learn more about the possibilities here, but have yet to do detailed reading on this.

**Comment by michaelweinold** [comment]
`(email by m-rossi to @michaelweinold and @cmutel from 24.08.2023)`

I am a little confused why you closed #616, If I read/understand everything correctly, no matter if we make bw2calc noarch or not there is an issue with the import of bwprocessing. If you fix this issue there it seems a recipe in emscripten-forge should be possible. This is still my preferred solution as we would make a package for the special wasm-usecase.

Regarding the noarch package of bw2calc on conda-forge:

Yes, it preferable to have a noarch package if it’s possible we should keep it noarch and not tell the users to manually install an additional package. Sadly optional packages are not possible with conda

A noarch package with non-noarch packages is possible. Problem we have on bw2calc is that scikit-umfpack is not available on Windows. If that would be the case we could make bw2calc noarch very easily

I am hesitant to add dev version which behaves differently than the main version, because at some point you add a new release version. What we do then?

I am unsure if the label approach is possible at all. You may give me some time researching this topic in the conda forge docs.

**Comment by cmutel** [comment]
> no matter if we make bw2calc noarch or not there is an issue with the import of bwprocessing

The `bw_processing`issue was solved via (a PR from @michaelweinold )[https://github.com/brightway-lca/bw_processing/pull/31], and a new version (0.8.4) was released and is already available on conda-forge.

> I am hesitant to add dev version which behaves differently than the main version, because at some point you add a new release version. What we do then?

I think as long as we keep the labels as separate branches than we are OK - we can update them independently, and the `main` branch can still be OS-specific. This might change in the future, there isn't a strict dependency for *any* linear algebra solver in `bw2calc`, it is just a convenience as many of our users are new to Python, so holding their hand can make for a better user experience.
